### PR TITLE
Errors moved from util to errors package  :seedling:

### LINKED ISSUE #5356: Move error variables out of utils package
**Detailed Description**

There are a few errors defined in the top level utils package that are better placed in the top level errors package. These errors should be deprecated and recreated in the errors package. The util errors should then be removed in a coming release.

Examples:
https://github.com/kubernetes-sigs/cluster-api/blob/573f005559b452813e4dd8175a791dc2d1b143ed/util/util.go#L58

https://github.com/kubernetes-sigs/cluster-api/blob/573f005559b452813e4dd8175a791dc2d1b143ed/util/util.go#L62

/kind cleanup

### PR DESCRIPTION
**Pull request by DiptoChakrabarty**
🌱

**What this PR does / why we need it**:
It moves the implementation of the  errors from util package to the error package 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5356 
(Working on it)

### CODE REVIEW

File: controllers/machine_controller_phases.go [hunk 1]
```diff
@@ -293,15 +293,15 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, cluster
 
 	// Get and set Status.Addresses from the infrastructure provider.
 	err = util.UnstructuredUnmarshalField(infraConfig, &m.Status.Addresses, "status", "addresses")
-	if err != nil && err != util.ErrUnstructuredFieldNotFound {
+	if err != nil && err != capierrors.ClusterFieldNotFound("field not found") {
```
Comment (enxebre) [inline controllers/machine_controller_phases.go:296]: why is this called ClusterFieldNotFound while fetching a Machine field?
Why is it changing the current implementation?

Comment (DiptoChakrabarty) [reply controllers/machine_controller_phases.go:296]: This is the name I had given to the function which returns a UnStructuredError error type , I can change the name if needed

Comment (killianmuldoon) [reply controllers/machine_controller_phases.go:296]: @DiptoChakrabarty This should be changed to capierrors.UnstructuredFieldNotFound

**Comment by killianmuldoon** [comment]
@DiptoChakrabarty on re-reviewing this I think we need to rethink the overall implementation - the errors package doesn't seem to fit as exactly the right place for these errors - the functions and values already in there are specifically those which appear in Status fields so the choices are:

**Comment by killianmuldoon** [comment]
@DiptoChakrabarty Thanks so much for your work on this so far! 🙌  I added a comment in the issue to get opinions on the overall approach we're introducing here so hopefully we can get some insight soon 😃

**Comment by vincepri** [comment]
Folks, just to chime in a bit here, I'm getting confused on the purpose of this overall change. In general, what we've done so far with errors is:

- If a package _returns_ an error that error must be defined in the same pacakge. 
- The error might be wrapped with errors.Wrapf(...)
- Callers that are checking the error from the function in package should use `errors.Is(...)` to check for the error and import the error from the other package.

The `capierrors` package was originally meant to be used to providers to create their own errors consistent with Cluster API, although it hasn't been a huge issue.

Let me know what you think

**Pull request closed without merging**